db/multi_cf_iterator.cc

```diff
@@ -9,54 +9,95 @@
 
 namespace ROCKSDB_NAMESPACE {
 
-void MultiCfIterator::SeekToFirst() {
+void MultiCfIterator::SeekCommon(
+    const std::function<void(Iterator*)>& child_seek_func,
+    Direction direction) {
+  direction_ = direction;
   Reset();
   int i = 0;
   for (auto& cfh_iter_pair : cfh_iter_pairs_) {
     auto& cfh = cfh_iter_pair.first;
     auto& iter = cfh_iter_pair.second;
-    iter->SeekToFirst();
+    child_seek_func(iter.get());
     if (iter->Valid()) {
       assert(iter->status().ok());
-      min_heap_.push(MultiCfIteratorInfo{iter.get(), cfh, i});
+      if (direction_ == kReverse) {
+        auto& max_heap = std::get<MultiCfMaxHeap>(heap_);
+        max_heap.push(MultiCfIteratorInfo{iter.get(), cfh, i});
+      } else {
+        auto& min_heap = std::get<MultiCfMinHeap>(heap_);
+        min_heap.push(MultiCfIteratorInfo{iter.get(), cfh, i});
+      }
     } else {
       considerStatus(iter->status());
     }
     ++i;
   }
 }
 
-void MultiCfIterator::Next() {
-  assert(Valid());
+template <typename BinaryHeap>
+void MultiCfIterator::AdvanceIterator(
+    BinaryHeap& heap, const std::function<void(Iterator*)>& advance_func) {
```

Comment:
Similarly to above, we can avoid the overhead of `std::function` by making the type of the functor another template parameter